### DISCUSSION TITLE
Help with my C++ parser for my Assembly language

### DISCUSSION THREAD

**Discussion by ghost** [community/community#120749] [Programming Help]
### Body

[The code is here](https://github.com/PyDOS8/BASICPY/blob/main/parser.cpp)

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CopperEagle** [comment]
## If you want a short answer and like templates
Writing a parser in C++ is not the easiest of things. Some languages like OCaml or Java have a much longer history in parsing and thus more tools and support to **generate** parsers with relatively little effort. Keyword: "Parser generator",

For C++, I have found [boost.spirit](https://boost-spirit.com/home/). It seems to be fairly advanced template stuff (read "hard-to-understand-errors"). Introductions seem rare but [here](https://www.codeproject.com/Articles/8516/An-Introduction-to-the-Boost-Spirit-Parser-framewo) is an old one and [here](https://panthema.net/2018/0912-Boost-Spirit-Tutorial/) one from KIT... but the video seems to be in german...
The other option is to write a parser from scratch.

## If you want to do it yourself
Upfront: If you have never written any parser you should probably start with a simpler language like arithmetics.

The answer is... you should look up the (E)BNF grammar and implement it one to one.
Parsing is one of these areas in practical Computer Science, which are pretty much just a bit of theory, the rest is can be automated.
(E)BNF grammars describe the syntax of the language to parse. It starts from the building blocks, the tokens (which you provided). Once you have the (E)BNF description for your language, you *are actually pretty much done*. There is even software that can generate a parser from the (E)BNF grammar. This section is about doing it manually.

Here is a small example to give you intuition how (E)BNF saves the game:

------------------------------------------------------------------------
\<oneArgFunction\> <== \<variable\> '(' \<expression\> ')' 
\<expression\> <== \<oneArgFunction\> | \<variable\> \<binop\> \<variable\> | ...
\<binop\>  <== "+" | "-" | "*" | "/"
------------------------------------------------------------------------

The way you parse this **top down**. Each function corresponds to one rule. Each function goes left to right through the rule and calls the correct function recursively, for each metasymbol (this \<blabla\>). This is how you would parse the above, in pseudocode.:

```py
# s is a stream ('list') of tokens
def parse_1_arg_function(s):
    fname = s.next_token(VARIABLE) # next_token throws error if not the case
    s.next_token( "(" )
    expr = parse_expression(s)
    s.next_token(")")
    return AstNode( FunctionCall( fname, expr ) )

def parse_expression(s):
    if s.is_next(parse_1_arg_function):
        return parse_1_arg_function(s)
    elif s.is_next(VARIABLE):
        var1 = parse_variable(s)
        binop = parse_binop(s)
        var2 = parse_variable(s)
        return AstNode( Binary(binop, var1, var2) )
    # ...
```

The above explanation is of course not exhaustive. There is some theory about grammars themselves and what can and cannot be done. But that's beyond the scope of a github comment...
